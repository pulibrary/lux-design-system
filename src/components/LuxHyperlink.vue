<template>
  <a
    :href="href"
    :class="['lux-link', variation, size, underline ? 'lux-link-underline' : '']"
    :target="newTab ? '_blank' : null"
  >
    <!-- @slot The text of your hyperlink. -->
    <slot />
    <lux-icon-base v-if="newTab" width="14" height="14" icon-name="(opens in new tab)">
      <lux-icon-new-tab></lux-icon-new-tab>
    </lux-icon-base>
  </a>
</template>

<script>
import LuxIconBase from "./icons/LuxIconBase.vue"
import LuxIconNewTab from "./icons/LuxIconNewTab.vue"

/**
 * Used to create hyperlinks as text or buttons. Can also be used on Card component
 * sub-elements to make the entire card click-able.
 */
export default {
  name: "LuxHyperlink",
  status: "ready",
  release: "1.0.0",
  type: "Element",
  props: {
    /**
     * The href value of the link.
     */
    href: {
      type: String,
      default: "",
      required: true,
    },
    /**
     * Whether the link appears as text or as a button. Options include `button solid`
     * and `button outline`.
     */
    variation: {
      type: String,
      default: "link",
      validator: value => {
        return value.match(/(link|button)/)
      },
    },
    /**
     * Sets the size of the button `small`, `medium`, `large`
     */
    size: {
      type: String,
      default: "medium",
      validator: value => {
        return value.match(/(small|medium|large)/)
      },
    },
    /**
     * Should the link open in a new tab?  This can be
     * disconcerting, so don't use it unless necessary.
     */
    newTab: {
      type: Boolean,
      default: false,
    },
    /**
     * Should the link have an underline?
     */
    underline: {
      type: Boolean,
      default: false,
    },
  },
  components: {
    LuxIconBase,
    LuxIconNewTab,
  },
}
</script>

<style lang="scss" scoped>
@import "../assets/styles/variables.css";
@import "../assets/styles/spacing.scss";

.lux-link {
  pointer-events: auto;
  color: var(--color-bleu-de-france-dark);
  font-family: var(--font-family-text);
  text-decoration: none;

  &:visited {
    color: var(--color-bleu-de-france-dark);
  }

  &:focus,
  &:hover {
    box-shadow: 0 2px currentColor;
  }

  &:active {
    box-shadow: 0 2px var(--color-bleu-de-france-darker);
  }

  &.button {
    line-height: var(--line-height-base);
    border: 0;
    border-radius: var(--border-radius-default);
    cursor: pointer;
    padding: 0.5rem 1rem;
    text-decoration: none;
    text-align: center;
    transition: background 250ms ease-in-out, transform 150ms ease;
    margin: 0 0.25rem;

    &:first-child {
      margin-left: 0;
    }

    &:last-child {
      margin-right: 0;
    }

    &.solid {
      background: var(--color-bleu-de-france);
      color: var(--color-white);
      &:hover,
      &:focus {
        background: var(--color-bleu-de-france-darker);
        box-shadow: none;
      }
    }

    &.outline {
      background: transparent;
      color: var(--color-bleu-de-france-dark);
      border: 2px solid var(--color-bleu-de-france);
      &:hover,
      &:focus {
        color: var(--color-bleu-de-france-darker);
        border-color: var(--color-bleu-de-france-darker);
        box-shadow: none;
      }
    }

    &.small {
      @include inset-space(12px);
      font-size: var(--font-size-small);
    }

    &.medium {
      @include inset-space(15px);
      font-size: var(--font-size-base);
    }

    &.large {
      @include inset-space(18px);
      font-size: var(--font-size-large);
    }
  }
}

.lux-link-underline {
  border-bottom: 2px black solid;

  &:focus,
  &:hover {
    border-bottom: 2px var(--color-princeton-orange-on-white) solid;
    box-shadow: none;
    text-decoration: none;
  }
}
</style>

<docs>
  ```jsx
    <div>
      <lux-hyperlink href="#">Foo</lux-hyperlink>
      <lux-hyperlink href="#" variation="button solid">Bar</lux-hyperlink>
      <lux-hyperlink href="#" variation="button solid" size="large">Bar</lux-hyperlink>
      <lux-hyperlink href="#" variation="button outline">Bar</lux-hyperlink>
      <lux-hyperlink href="#" :newTab="true">I open in a new tab</lux-hyperlink>
      <lux-hyperlink href="#" :underline="true">
        I am underlined and have an arrow
        <lux-icon-base width="14" height="14">
          <lux-icon-arrow-right></lux-icon-arrow-right>
        </lux-icon-base>
      </lux-hyperlink>
    </div>
  ```
</docs>
