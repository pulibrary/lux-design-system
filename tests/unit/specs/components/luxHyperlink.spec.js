import { mount } from "@vue/test-utils"
import LuxHyperlink from "@/components/LuxHyperlink.vue"
import { nextTick } from "vue"

describe("LuxHyperlink.vue", () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(LuxHyperlink, {
      props: {
        href: "https://library.princeton.edu",
      },
      slots: {
        default: "I'll take you places",
      },
    })
  })

  it("should render the appropriate level heading", () => {
    wrapper.setProps({ href: "https://library.princeton.edu" })
    const link = wrapper.find("a")
    expect(link.text()).toBe("I'll take you places")
  })

  it("should render the appropriate class name", async () => {
    wrapper.setProps({ href: "https://library.princeton.edu", variation: "button" })
    await nextTick()
    const link = wrapper.find("a")
    expect(link.classes()).toContain("button")
  })

  it("does not have a target attribute by default", () => {
    expect(wrapper.find("a").attributes("target")).toBeUndefined()
  })

  it("does not have an icon by default", () => {
    expect(wrapper.find("a").find("svg").exists()).toBe(false)
  })

  describe("with the newTab prop", () => {
    beforeEach(async () => {
      wrapper.setProps({ newTab: true })
      await nextTick()
    })

    it("adds target _blank", () => {
      expect(wrapper.find("a").attributes("target")).toEqual("_blank")
    })

    it("adds an icon", () => {
      const link = wrapper.find("a")
      expect(link.find("svg").exists()).toBe(true)
    })

    it("tells screen readers that the link opens in a new tab", () => {
      expect(wrapper.find("a").text()).toContain("opens in new tab")
    })
  })

  describe("with the underline prop", () => {
    beforeEach(async () => {
      wrapper.setProps({ underline: true })
      await nextTick()
    })

    it("adds an underline class", () => {
      expect(wrapper.find("a").classes()).toContain("lux-link-underline")
    })
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
