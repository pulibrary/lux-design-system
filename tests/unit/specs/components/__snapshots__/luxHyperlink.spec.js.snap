// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuxHyperlink.vue has the expected html structure 1`] = `
<a
  class="lux-link link medium"
  href="https://library.princeton.edu"
>
  <!-- @slot The text of your hyperlink. -->
  
  I'll take you places
  
  <!--v-if-->
</a>
`;
